File: examples/lua/kademlia/routing_table.lua
-- Routing table data
routing_table = {
buckets = {},
}
-- Initialize the routing table
function routing_table_init(id) 
	routing_table.id = id
	-- Bucket initialization
	for i = 0,common.IDENTIFIER_SIZE do
		routing_table.buckets[i] = {id = i, set = {}, list = {}}
	end
end
-- Returns an identifier which is in a specific bucket of a routing table
function get_id_in_prefix(id, prefix)
	if id == 0 then
		return 0
	end
	local identifier = 1
	identifier = math.pow(identifier,prefix - 1)
	identifier = bxor(identifier,id)
	return identifier
end
-- Returns the corresponding node prefix for a given id
function get_node_prefix(id)
	for i = 0,32 do
		if is_integer(id / math.pow(2,i)) and (id / math.pow(2,i)) % 2 == 1 then
			return 31 - i
		end
	end 
	return 0
end
-- Finds the corresponding bucket in a routing table for a given identifier
function find_bucket(id)
	local xor_number = bxor(id,routing_table.id)
	local prefix = get_node_prefix(xor_number)
	--simgrid.info("Prefix:" .. prefix .. " number:" .. xor_number)
	return routing_table.buckets[prefix]
end
-- Updates the routing table with a new value.
function routing_table_update(id)
	if id == routing_table.id then
		return
	end
	local bucket = find_bucket(id)
	if bucket.set[id] ~= nil then
		simgrid.debug("Updating " .. id .. " in my routing table")
		-- If the element is already in the bucket, we update it.
		table.remove(bucket.list,index_of(bucket.list,id))
		table.insert(bucket.list,0,id)
	else	
		simgrid.debug("Insert " .. id .. " in my routing table in bucket " .. bucket.id)
		table.insert(bucket.list,id)
		bucket.set[id] = true
	end
end
-- Returns the closest notes we know to a given id 
function find_closest(destination_id)
	local answer = {destination = destination_id, nodes = {}}
	
	local bucket = find_bucket(destination_id)
	for i,v in pairs(bucket.list) do
		table.insert(answer.nodes,{id = v, distance = bxor(v,destination_id)})
	end
	
	local i = 1
	
	while #answer.nodes < common.BUCKET_SIZE and ((bucket.id - i) >= 0 or (bucket.id + i) <= common.IDENTIFIER_SIZE) do
		-- Check the previous buckets
		if bucket.id - i >= 0 then
			local bucket_p = routing_table.buckets[bucket.id - i]
			for i,v in pairs(bucket_p.list) do
				table.insert(answer.nodes,{id = v, distance = bxor(v,destination_id)})
			end
		end			
		-- Check the next buckets
		if bucket.id + i <= common.IDENTIFIER_SIZE then
			local bucket_n = routing_table.buckets[bucket.id + i]
			for i,v in pairs(bucket_n.list) do
				table.insert(answer.nodes,{id = v, distance = bxor(v,destination_id)})
			end			
		end
		i = i + 1
	end
	-- Sort the list
	table.sort(answer.nodes, function(a,b) return a.distance < b.distance end)
	-- Trim the list
	while #answer.nodes > common.BUCKET_SIZE do
		table.remove(answer.nodes)
	end
	
	return answer
end

File: examples/java/kademlia/RoutingTable.java
/* Copyright (c) 2012-2014, 2016. The SimGrid Team.
 * All rights reserved.                                                     */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

package kademlia;
import java.util.Collections;
import java.util.Vector;

import org.simgrid.msg.Msg;

public class RoutingTable {
  /* Bucket list */
  private Vector<Bucket> buckets;
  /* Id of the routing table owner */
  private int id;

  public RoutingTable(int id) {
    this.id = id;
    buckets = new Vector<Bucket>();
    for (int i = 0; i < Common.IDENTIFIER_SIZE + 1; i++) {
      buckets.add(new Bucket(i));
    }
  }

  /**
   * @brief Returns an identifier which is in a specific bucket of a routing table
   * @param id id of the routing table owner
   * @param prefix id of the bucket where we want that identifier to be
   */
  public int getIdInPrefix(int id, int prefix) {
    if (prefix == 0) {
      return 0;
    }
    int identifier = 1;
    identifier = identifier << (prefix - 1);
    identifier = identifier ^ id;
    return identifier;
  }

  /* Returns the corresponding node prefix for a given id */
  public int getNodePrefix(int id) {
    for (int j = 0; j < 32; j++) {
      if ((id >> (32 - 1 - j) & 0x1) != 0) {
        return 32 - j;
      }
    }
    return 0;
  }

  /* Finds the corresponding bucket in a routing table for a given identifier */
  public Bucket findBucket(int id) {
    int xorNumber = id ^ this.id;
    int prefix = this.getNodePrefix(xorNumber);
    return buckets.get(prefix);
  }

  /* Updates the routing table with a new value. */
  public void update(int id) {
    Bucket bucket = this.findBucket(id);
    if (bucket.contains(id)) {
      Msg.debug("Updating " + Integer.toString(id) + " in my routing table");
      //If the element is already in the bucket, we update it.
      bucket.pushToFront(id);
    } else {
      Msg.debug("Adding " + id + " to my routing table");
      bucket.add(id);
      if (bucket.size() > Common.BUCKET_SIZE)  {
        //TODO: Ping the least seen guy and remove him if he is offline.
      }
    }
  }

  /* Returns the closest notes we know to a given id */
  public Answer findClosest(int destinationId) {
    Answer answer = new Answer(destinationId);
    Bucket bucket = this.findBucket(destinationId);
    bucket.addToAnswer(answer,destinationId);

    for (int i = 1; answer.size() < Common.BUCKET_SIZE && ((bucket.getId() - i) >= 0 ||
                                    (bucket.getId() + i) <= Common.IDENTIFIER_SIZE); i++) {
      //Check the previous buckets
      if (bucket.getId() - i >= 0) {
        Bucket bucketP = this.buckets.get(bucket.getId() - i);
        bucketP.addToAnswer(answer,destinationId);
      }
      //Check the next buckets
      if (bucket.getId() + i <= Common.IDENTIFIER_SIZE) {
        Bucket bucketN = this.buckets.get(bucket.getId() + i);
        bucketN.addToAnswer(answer, destinationId);
      }
    }
    //We sort the list
    Collections.sort(answer.getNodes());
    //We trim the list
    while (answer.size() > Common.BUCKET_SIZE) {
      answer.remove(answer.size() - 1); //TODO: Not the best thing.
    }
    return answer;
  }

  @Override
  public String toString() {
    String string = "RoutingTable [ id=" + id + " " ;
    for (int i = 0; i < buckets.size(); i++) {
      if (buckets.get(i).size() > 0) {
        string += buckets.get(i) + " ";
      }
    }
    return string;
  }
}
